### ISSUE TITLE
[discussion] Change Eigen to have no dependencies?

### ISSUE THREAD

**Issue by svenevs** [spack/spack#7554]
Why are there any dependencies for Eigen at all?  It's a header only library and the installed CMake scripts do not encode the backend choices (dependencies) that are currently used.

### The Installation

```cmake
set (EIGEN3_DEFINITIONS  "")
set (EIGEN3_INCLUDE_DIR  "${PACKAGE_PREFIX_DIR}/include/eigen3")
set (EIGEN3_INCLUDE_DIRS "${PACKAGE_PREFIX_DIR}/include/eigen3")
set (EIGEN3_ROOT_DIR     "${PACKAGE_PREFIX_DIR}")
```

`EIGEN3_DEFINITIONS` (as far as I know) would be for setting something like `EIGEN_MAX_ALIGN_BYTES=0` (e.g. for Win32).  But in reality, I rarely see any build systems that use Eigen even reference `EIGEN3_DEFINITIONS`.

When a user of Eigen wants to enable a specific backend, they must find / include / link with it themselves.  For example, [the FFT backend](http://eigen.tuxfamily.org/index.php?title=EigenFFT):

> - compiling code with preprocessor definition `EIGEN_FFTW_DEFAULT`
> - linking with FFTW libraries e.g. `-lfftw3 -lfftw3f -lfftw3l`

Even if it were possible, Spack shouldn't change this behavior either (by somehow enabling `EIGEN_FFTW_DEFAULT` or linking flags in the installed CMake / pkg-config scripts).  I've always *interpreted* this setup as "pay for what you use" at it's beautiful C++ core.

Since it's a header only library, Eigen places tuning / optimization / backend choice responsibilities on the user.  For example, [the user is responsible for compilation flags related to vectorization](http://eigen.tuxfamily.org/index.php?title=FAQ#How_can_I_enable_vectorization.3F).

### Why Keep the Spack Dependencies

Projects that depend on Eigen and **also** use / enable a specific backend will always have it available.  For example, `fftw` will currently be installed and available for any dependent of Eigen during the build phase.

### Why Get Rid of the Dependencies

It takes a long time to install a header only library, and currently not all possible backends are even encoded in Spack.  For example, `superlu` and `intel-mkl` could also be variants for [external solvers](https://eigen.tuxfamily.org/dox/group__TopicSparseSystems.html#title3).

It would take some doing and testing, but I think the backend choices should be removed entirely from Eigen.  The responsibility should be on the package that depends on Eigen instead.  For example the `vpfft` package also currently depends on `fftw`.  It doesn't look like that package actually uses the Eigen FFT code, but if it did then that package would be responsible.

In practice, if a library is using a specific back end, their build system would already be setup to search for / use it.

Basically, to remove them, we would need to inspect the current `dependents` of the `eigen` package and make sure that they are going to work correctly still.  I am more than willing to get work started on this, but ideally people familiar with a given package could also review.

**Comment by adamjstewart** [comment]
To back up your argument, [Homebrew's eigen package](https://github.com/Homebrew/homebrew-core/blob/master/Formula/eigen.rb) also doesn't have any dependencies aside from CMake.

**Comment by adamjstewart** [comment]
If I'm reading it correctly, [Easybuild's eigen package](https://github.com/easybuilders/easybuild-easyblocks/blob/master/easybuild/easyblocks/e/eigen.py) has no dependencies at all (unless they are defined elsewhere, I'm not that familiar with how Easybuild does things).

**Comment by adamjstewart** [comment]
It looks like @alalazo created this package. Any comments on the need for these dependencies?

**Comment by alalazo** [comment]
That was quite some time ago, but I remember that back then I inspected the `CMakeLists.txt` and looked at what Eigen was looking for in terms of `find_package` requests. I kind of expected that at least some macros got defined somewhere (like a header file) depending on the CMake configuration. To be honest I never double checked if that was really the case.

Bottomline: if it seems the right thing to do, please remove the dependencies. :grin: 

p.s. In case it could be nice to have `headers` and `libs` properties in Eigen that return the correct set of macros or link flags depending on how people want to configure the library in their projects.

**Comment by svenevs** [comment]
Mmmmmk I'll get started later today.

> In case it could be nice to have headers and libs properties in Eigen that return the correct set of macros or link flags depending on how people want to configure the library in their projects.

Can you elaborate on this?

**Comment by alalazo** [comment]
According to [this page](http://eigen.tuxfamily.org/dox/TopicUsingBlasLapack.html), or similar ones in Eigen's documentation, the user needs to define some macros to enable the use of external libraries. I think it makes sense for an application using e.g. `eigen` and `intel-mkl` to be able to say:
```python
eigen_macros = self.spec['eigen:intel-mkl'].headers.macro_definitions
```
to get the right set of macros for the case they are dealing with.

**Comment by svenevs** [comment]
In order to do that, every possible backend would then be a `variant`.  But I still think that would deceive future packagers.  I guess the point I'm trying to make is that the package build system itself must be responsible for doing this.  Even if `spack` were enabling the correct definitions

1. Adding those definitions may conflict with the packages build system.
2. If the definitions are added, but the package build system is not finding / including / linking with BLAS, the package build system now must be patched -- enabling the preprocessor definition but nothing else will lead to compilation errors.

One example is OpenCV.  Even though the OpenCV build system does things like detecting OpenBLAS and MKL, as far as I can tell [their treatment of Eigen is "vanilla"](https://github.com/opencv/opencv/blob/98b6e1d5fec584b27463dfcab3fef6c0a1cf92ed/cmake/OpenCVFindLibsPerf.cmake#L47).  Part of the reason for this is no Eigen types are exposed as part of the public API.  At least I think?

**To be clear**, I'm not actually disagreeing here.  I think what you're talking about would be a nice feature, I just don't know how to approach it so that things don't break.  I wish I could ask on the Eigen lists, but they've rejected me joining for over a year now :cry:

**Comment by alalazo** [comment]
> In order to do that, every possible backend would then be a variant

`eigen` should know what to answer when asked:
```python
eigen_macros = self.spec['eigen:intel-mkl'].headers.macro_definitions
```
but that doesn't mean it needs variants. It just needs to support the right set of query parameters via its "build interface". If you look at `openmpi`, for instance, it supports:
```python
spec['openmpi:cxx'].libs
```
even though `cxx` is not a variant.

> Adding those definitions may conflict with the packages build system.

This part I don't get. If we have an application that uses `eigen`, and its build system is well written, chances are that we don't need to ask any additional information to `eigen`. Usually build interfaces are most useful for `Makefile` based packages, of for those packages that leaves up to the user certain parts of the linking and compilation process. One prominent example of this might be `cp2k`.

Also I think the two modifications (remove dependencies from eigen + add support for build interface queries) are separate and can be subject of different PRs. What I wanted to say above is just that the information [here](https://eigen.tuxfamily.org/dox/TopicUsingBlasLapack.html) can probably be encoded in `eigen`'s build interface.

**Comment by svenevs** [comment]
Ok, the Eigen changes are going to take a while to get through anyway though.  I've only gone through OpenCV (and determined no deps will work as expected).  Assuming everything is good to go without this fancy stuff, I'll submit a PR with that.

-------------------------------------------

As for conflicting definitions, I think it's only a concern for things like enabling / disabling alignment and vectorization.  I don't think it would be an issue for back end choices, I thought it would be for MKL but `EIGEN_USE_MKL_ALL`  defines `EIGEN_USE_BLAS`, `EIGEN_USE_LAPACKE`, and `EIGEN_USE_MKL_VML` (I thought `EIGEN_USE_BLAS` would be the conflict, but that was dumb).

The main concern for me is that if we add `EIGEN_USE_BLAS`, who is responsible for finding the headers and libraries associated with this?  The secondary package that depends on `eigen`, or that package's build system?

- FooPackage is responsible: build system most likely has to be patched to find / use blas.
- Foo build system is responsible: most likely they're already defining `EIGEN_USE_BLAS` **if** using the BLAS backend has any meaning for the package.

**Comment by svenevs** [comment]
I'm not going to be fixing this, I stopped using spack to install eigen in favor of external installs.  Works well enough.
